Public/Src/Utilities/Native/IO/Helpers.cs

```diff
@@ -135,12 +138,76 @@ public static class Helpers
                     // Only log exceptions if they are not being re-thrown already
                     if (logExceptions)
                     {
-                        Tracing.Logger.Log.RetryOnFailureException(Events.StaticContext, e.Message ?? e.ToString());
+                        Tracing.Logger.Log.RetryOnFailureException(Events.StaticContext, e.ToString());
                     }
                 }
             }
 
             return done;
         }
+
+        /// <summary>
+        /// A generic version of <see cref="RetryOnFailure"/>.
+        /// </summary>
+        public static Possible<TResult, Failure> RetryOnFailure<TResult>(
+          Func<bool, Possible<TResult, Failure>> work,
+          bool logExceptions = true,
+          bool rethrowExceptions = false,
+          int numberOfAttempts = DefaultNumberOfAttempts,
+          int initialTimeoutMs = DefaultInitialTimeoutMs,
+          int postTimeoutMultiplier = DefaultPostTimeoutMultiplier)
+        {
+            Possible<TResult, Failure>? possiblyResult = null;
+
+            bool success = RetryOnFailure(
+                isLastRetry =>
+                {
+                    possiblyResult = work(isLastRetry);
+                    return possiblyResult.Value.Succeeded;
+                },
+                logExceptions, rethrowExceptions, numberOfAttempts, initialTimeoutMs, postTimeoutMultiplier);
+
+            // Since work returns Possible<TResult, Failure>, we expect it to gracefully handle any exception
+            // (i.e., return a failure instead of throwing an exception). However, there is no guarantee that
+            // it would behave this way. 
+            if (!possiblyResult.HasValue)
+            {                
+                Contract.Assert(false, $"An operation resulted in an unhandled exception in all of the '{numberOfAttempts}' retry attempts");
+            }
+
+            return possiblyResult.Value;
+        }

```

Comment:
You don't need these two methods. See method `TryMaterializeAsync` in `Public\Src\Engine\Cache.Plugin.CacheCore\CacheCoreArtifactContentCache.cs`